platform/src/server/http/http_server.ts

```diff
@@ -1,51 +1,82 @@
-import express from 'express';
-import http from 'http';
-import https from 'https';
-import { Server } from 'net';
-import { promisify } from 'util';
+import { Request, ResponseToolkit, Server, ServerOptions } from 'hapi';
+import { ServerOptions as TLSOptions} from 'https';
 import { readFileSync } from 'fs';
 
 import { HttpConfig } from './http_config';
 import { Env } from '../../config';
 import { Logger } from '../../logging';
-import { Router } from './Router';
+import { Router } from './router';
+import { modifyUrl } from '../../lib/utils/url';
 
 export class HttpServer {
-  private readonly app: express.Application;
   private server?: Server;
+  private registeredRouters: Set<Router> = new Set();
 
   constructor(private readonly log: Logger, private readonly env: Env) {
-    this.app = express();
   }
 
   isListening() {
-    return this.server !== undefined && (this.server as any).listening;
+    return this.server !== undefined && this.server.listener.listening;
   }
 
   registerRouter(router: Router) {
-    this.app.use(router.path, router.router);
+    if (this.isListening()) {
+      throw new Error('Routers can be registered only when HTTP server is stopped.')
+    }
+
+    this.registeredRouters.add(router);
   }
 
   async start(config: HttpConfig) {
-    const server = this.initializeServer(config);
-
-    this.server = server;
+    this.server = this.initializeServer(config);
+
+    this.setupBasePathRewrite(this.server, config);
+
+    for (const router of this.registeredRouters) {
+      for (const route of router.getRoutes()) {
+        this.server.route({
+          method: route.method,
+          path: `${router.path}${route.path}`,
+          handler: route.handler,
+        });
```

Comment:
There's a gap in features here between new platform and old. I'm not sure if that's a problem we need to address in this PR, but I wanted to raise it. We currently use hapi's `options` hash to specify things like validation, auth, and tags. How do you suppose we should handle these things in the new platform going forward?

Reply:
> There's a gap in features here between new platform and old. I'm not sure if that's a problem we need to address in this PR, but I wanted to raise it. We currently use hapi's options hash to specify things like validation, auth, and tags. How do you suppose we should handle these things in the new platform going forward?

Yeah, there is a gap, but I'd define first what we want to support in new platform and why we can't live without it :) Some of the Hapi "features" are just handled differently in new platform. E.g.:

* Validation - we have some support for this already, see https://github.com/elastic/kibana/blob/new-platform/example_plugins/baz/src/register_endpoints.ts#L33-L48. Whether it's sufficient or not is a different story.

* Auth - we didn't discuss this a lot, but it should be solely handled by security plugin. It will also expose API that other plugins can consume (e.g. to get current user, or mark the route as one that doesn't require authentication or something like this)

* Tags - why do we need to support this in new platform? I can see that route metadata can be helpful (e.g. to have self-documented routes), but can't think of anything critical that can't implemented in the next iteration.

Reply:
We do use route tags at the moment. Whether or not we want to expose that same concept in the platform is probably a separate discussion.

We currently use route tags to flag whether an endpoint should respond with a redirect to login or a 403. I'm not sure if this is even necessary at this point, but it's functionality that currently exists.

Reply:
Yeah, it's used in security plugin, we'll figure that out in the scope of https://github.com/elastic/kibana/issues/18024